pkg/redirectpolicy/manager_test.go

```diff
@@ -0,0 +1,432 @@
+		id: k8s.ServiceID{
+			Name:      "test-foo",
+			Namespace: "ns1",
+		},
+		lrpType: lrpConfigTypeSvc,
+		backendSelector: api.EndpointSelector{
+			LabelSelector: &slim_metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"test": "foo",
+				},
+			},
+		},
+	}
+}
+
+// Tests if duplicate addressMatcher configs are not added.
+func (m *ManagerSuite) TestManager_AddRedirectPolicy_AddrMatcherDuplicateConfig(c *C) {
+	configFe := configAddrType
+	m.rpm.policyFrontendsByHash[fe1.Hash()] = configFe.id
+	dupConfigFe := configFe
+	dupConfigFe.id.Name = "test-foo2"
+
+	added, err := m.rpm.AddRedirectPolicy(dupConfigFe, nil)
+
+	c.Assert(added, Equals, false)
+	c.Assert(err, NotNil)
+}
+
+// Tests if duplicate svcMatcher configs are not added.
+func (m *ManagerSuite) TestManager_AddRedirectPolicy_SvcMatcherDuplicateConfig(c *C) {
+	configSvc := configSvcType
+	configSvc.serviceID = &k8s.ServiceID{
+		Name:      "foo",
+		Namespace: "ns1",
+	}
+	m.rpm.policyConfigs[configSvc.id] = &configSvc
+	m.rpm.policyServices[*configSvc.serviceID] = configSvc.id
+	invalidConfigSvc := configSvc
+	invalidConfigSvc.id.Name = "test-foo3"
+
+	added, err := m.rpm.AddRedirectPolicy(invalidConfigSvc, nil)
+
+	c.Assert(added, Equals, false)
+	c.Assert(err, NotNil)
+}
+
+// Tests add redirect policy, add pod, delete pod and delete redirect policy events
+// for an addressMatcher config with a frontend having single port.
+func (m *ManagerSuite) TestManager_AddrMatcherConfigSinglePort(c *C) {
+	// Add an addressMatcher type LRP with single port. The policy config
+	// frontend should have 2 pod backends with each of the podIPs.
+	configAddrType.frontendType = addrFrontendSinglePort
+	configAddrType.backendPorts = append(configAddrType.backendPorts, beP1)
+	podIPs, _ := utils.ValidIPs(pod1.Status)
+	expectedbes := make([]backend, len(podIPs))
+	for i := range podIPs {
+		expectedbes[i] = backend{
+			L3n4Addr: lb.L3n4Addr{IP: net.ParseIP(podIPs[i]), L4Addr: beP1.l4Addr},
+			podID:    pod1ID,
+		}
+	}
+
+	added, err := m.rpm.AddRedirectPolicy(configAddrType, &fakePodStore{})
+
+	c.Assert(added, Equals, true)
+	c.Assert(err, IsNil)
+	c.Assert(len(m.rpm.policyConfigs), Equals, 1)
+	c.Assert(m.rpm.policyConfigs[configAddrType.id].id.Name, Equals, configAddrType.id.Name)
+	c.Assert(m.rpm.policyConfigs[configAddrType.id].id.Namespace, Equals, configAddrType.id.Namespace)
+	c.Assert(len(m.rpm.policyFrontendsByHash), Equals, 1)
+	c.Assert(m.rpm.policyFrontendsByHash[configAddrType.frontendMappings[0].feAddr.Hash()],
+		Equals, configAddrType.id)
+	c.Assert(len(configAddrType.frontendMappings[0].podBackends), Equals, 2)
+	for i := range configAddrType.frontendMappings[0].podBackends {
+		c.Assert(configAddrType.frontendMappings[0].podBackends[i], checker.Equals, expectedbes[i])
+	}
+	c.Assert(len(m.rpm.policyPods), Equals, 1)
+	c.Assert(len(m.rpm.policyPods[pod1ID]), Equals, 1)
+	c.Assert(m.rpm.policyPods[pod1ID][0], Equals, configAddrType.id)
```

Comment:
I'm glad to see unit tests, though I would prefer if we took a more black-box approach so that the unit tests aren't so tied into the internal implementation but rather to the interfaces that the object interacts with.

What I mean is, rather than doing:

* AddRedirectPolicy
* Check internal ap1 is updated
* Check internal map2 is updated
* Check other internal state
* ...

Instead:

* AddRedirectPolicy
* Check that the podStore, servicemanager get the correct number of lookups/updates with the correct contents.
* Add another pod
* Check that the servicemanager gets frontend updates that reflect the pod change
* ...

With this approach, in future we can modify the internal representation of the data to suit a different implementation, and the unit tests will still work and won't need changing when we change the implementation.

Given that we've just merged this implementation and probably won't go and change it drastically any time soon, I'm not *so* worried to address this in this PR, but this would be good to keep in mind in the future. The reason I bring this up is that there has been other test code written in Cilium in the past which is directly tied to the internals like this, introspecting internal maps and whatnot. When we go to refactor that code, it can be painful to go through each test one-by-one to update the references to refer to the new implementation.

Reply:
I agree with checking the number of calls. In fact, while testing the last fix, I found an issue that wasn't flagged by the unit tests, and made a mental note about this. I'll enhance the unit tests approach.